engine/src/flutter/display_list/dl_canvas.cc

```diff
@@ -4,22 +4,132 @@
 
 #include "flutter/display_list/dl_canvas.h"
 
-#include "flutter/display_list/geometry/dl_geometry_conversions.h"
-#include "flutter/third_party/skia/include/core/SkPoint3.h"
-#include "flutter/third_party/skia/include/utils/SkShadowUtils.h"
+namespace {
+
+// ShadowBounds code adapted from SkShadowUtils using the Directional flag.
+
+using DlScalar = flutter::DlScalar;
+using DlVector3 = flutter::DlVector3;
+using DlVector2 = flutter::DlVector2;
+using DlRect = flutter::DlRect;
+using DlMatrix = flutter::DlMatrix;
+
+static constexpr DlScalar kAmbientHeightFactor = 1.0f / 128.0f;
+static constexpr DlScalar kAmbientGeomFactor = 64.0f;
+// Assuming that we have a light height of 600 for the spot shadow, the spot
+// values will reach their maximum at a height of approximately 292.3077.
+// We'll round up to 300 to keep it simple.
+static constexpr DlScalar kMaxAmbientRadius =
+    300.0f * kAmbientHeightFactor * kAmbientGeomFactor;
+
+inline DlScalar AmbientBlurRadius(DlScalar height) {
+  return std::min(height * kAmbientHeightFactor * kAmbientGeomFactor,
+                  kMaxAmbientRadius);
+}
+
+struct DrawShadowRec {
+  DlVector3 light_position;
+  DlScalar light_radius = 0.0f;
+  DlScalar occluder_z = 0.0f;
+};
+
+static inline float DivideAndClamp(float numer,
+                                   float denom,
+                                   float min,
+                                   float max) {
+  float result = std::clamp(numer / denom, min, max);
+  // ensure that clamp handled non-finites correctly
+  FML_DCHECK(result >= min && result <= max);
+  return result;
+}
+
+inline void GetDirectionalParams(DrawShadowRec params,
+                                 DlScalar* blur_radius,
+                                 DlScalar* scale,
+                                 DlVector2* translate) {
+  *blur_radius = params.light_radius * params.occluder_z;
+  *scale = 1.0f;
+  // Max z-ratio is ("max expected elevation" / "min allowable z").
+  constexpr DlScalar kMaxZRatio = 64.0f / flutter::kEhCloseEnough;
+  DlScalar zRatio = DivideAndClamp(params.occluder_z, params.light_position.z,
+                                   0.0f, kMaxZRatio);
+  *translate = DlVector2(-zRatio * params.light_position.x,
+                         -zRatio * params.light_position.y);
+}
+
+DlRect GetLocalBounds(DlRect ambient_bounds,
```

Comment:
Can you add a docstring here.  It's not immediately easy to understand what this is doing.  Here is the one for skia: https://github.com/google/skia/blob/d43c7d1ea21a80982b926be4f72f48b64fa4bfa6/include/utils/SkShadowUtils.h#L66

The name isn't quite obvious like the skia version either IMO since it's laking the "shadow" context.

Reply:
Well, that's the implementation function which never has a doc string. The doc string should be in the header file and that didn't change so it isn't in the diffs.

But, if you go look there you'll notice that it is also missing a doc comment, so the point is still valid. ;)

Reply:
None of the methods in DlCanvas are documented as they are assumed to be just duplicates of the SkCanvas methods, but we need to fix that at some point. For now, I'm also adding a doc comment to DrawShadow as well as the ComputeBounds function to help fully document what we mean by shadows (ComputeBounds refers to DrawShadow as "I'm the bounds for that guy" so it helps to nail down what "that guy" does). The rest of the methods in DlCanvas should be documented later as a low priority task in the de-skiafication mini-project.